[Reservation] 예약 기능 정보 저장

J-Cheol opened the pull request:
- close #37 

### 작업목록
- [x] Reservation Entity 생성
- [x] ReservationProductAddOption Entity 생성
- [x] 해당 Repository 생성
- [x] 예약 정보 저장 비즈니스 로직 생성

### 참고사항
현재 LocalTime이 HH:mm으로 들어가고 있지 않습니다.
Json포멧을 변경하여도 계속 HH:mm:ss로 DB에 저장되고 있어서 그대로두었습니다.
조회 만들때 수정하거나 로직을 변경하겠습니다.

J-Cheol committed:
feat : 예약 기능 연관관계 설정

ProductAddOption Entity - id타입 Long으로 변경

J-Cheol committed:
feat : 예약 정보 저장 기능을 위한 repository 생성

J-Cheol committed:
feat : 예약 정보 저장 비즈니스 로직 생성

J-Cheol committed:
feat : Swagger 예약 정보 저장 API 작성

J-Cheol committed:
feat : LocalTime createTime 생성 및 로직 적용

J-Cheol committed:
feat : status enum 구현

ppusda commented on src/main/java/com/toucheese/reservation/service/ReservationService.java:
```diff
@@ -0,0 +1,67 @@
+			})
+			.collect(Collectors.toList());
+
+		Reservation reservation = new Reservation(
```
요건 `@Builder`로 해도 좋을 것 같습니다!

J-Cheol replied:
entity @builder 및 AllArgsConstructor 생성 및 builder로 변경하였습니다.

ppusda commented on src/main/java/com/toucheese/reservation/service/ReservationService.java:
```diff
@@ -0,0 +1,67 @@
+
+	@Transactional
+	public Reservation createReservation(ReservationRequest reservationRequest) {
+		Product product = productRepository.findById(reservationRequest.productId())
```
타 도메인의 리포지토리를 사용하는 부분은 지양하는 게 좋을 것 같습니다.

각 서비스에 `findXXXById`를 만들어두고 Service 관계로 연결하여 재사용성을 늘리고 관심사를 분리할 수 있을 것 같습니다.

J-Cheol replied:
감사합니다. 알려주신대로 바꾸는게 더 깔끔하고 재사용성도 좋네요!

ppusda commented on src/main/java/com/toucheese/reservation/service/ReservationService.java:
```diff
@@ -0,0 +1,67 @@
+					.orElseThrow(() -> new ToucheeseBadRequestException("AddOption not found"));
+				if (!product.getProductAddOptions().contains(productAddOption)) {
+					throw new ToucheeseBadRequestException("Invalid AddOption for the selected product");
+				}
```
분기문 내부에서 throw를 던지는 경우는 `validateAddOption`과 같이 함수로 분리 할 수 있을 것 같습니다!

J-Cheol replied:
왜 이건 제가 분리를 안해놨었는지 이해가 안되네요 ㅋㅋㅋㅋ 머쓱 감사합니다.

ppusda commented on src/main/java/com/toucheese/reservation/entity/Reservation.java:
```diff
@@ -0,0 +1,86 @@
+	@Enumerated(EnumType.STRING)
+	private ReservationStatus status;
+
+	@PrePersist
```
PrePersist는 처음 보네용
초기 값 설정을 위해 사용하신건가요?

생성자에서 할당하는 방법하고는 어떻게 다른가요?

J-Cheol replied:
PrePersist는 초기 값 설정을 위해 사용하였습니다.
PrePersist - DB에 저장되기 직전에 초기화를 해주어 혹여 다른 값이 status에 들어간 상태라고 해도 저장되기 전에 초기화를 시켜줍니다. 실수로 객체 생성시 null값 혹은 다른 값을 넣어도 해당 값으로 초기화됩니다.

ppusda commented on src/main/java/com/toucheese/reservation/controller/ReservationController.java:
```diff
@@ -0,0 +1,87 @@
+	public ResponseEntity<ReservationRequest> reservationCreate(@RequestBody ReservationRequest reservationRequest) {
+		String name = reservationRequest.name();
+		String phone = reservationRequest.phone();
+		if (name.isBlank() || phone.length() < 10 || phone.length() > 13) {
```
`@Valid`를 사용하지 않고 직접 검증을 진행하신 이유가 있으실까요?

J-Cheol replied:
@Valid를 생각하지 못하였네요.
비회원으로 구상했을때라 단순하게 값이 들어왔을때 이름과 핸드폰번호 자리수만 생각해서 처리했었습니다. 수정완료

ppusda commented on src/main/java/com/toucheese/reservation/controller/ReservationController.java:
```diff
@@ -0,0 +1,87 @@
+	@PostMapping("")
+	@Operation(
+		summary = "예약 정보 저장 기능",
+		description = "필요한 예약 정보 : <br>" +
```
지금 보니 """로 하시면 +를 사용하지 않으셔도 되었을 것 같네요... ㅠ

J-Cheol replied:
이런... 엔터를 누르니까 자동으로 생겨서 그러려니 했는데 감사합니다.

ppusda commented on src/main/java/com/toucheese/reservation/controller/ReservationController.java:
```diff
@@ -0,0 +1,87 @@
+		}
+		Reservation reservation = reservationService.createReservation(reservationRequest);
+
+		Product product = productRepository.findById(reservationRequest.productId())
```
Controller <-> Service <-> Repository 계층별로 분리가 필요해보입니다.
아래 내용을 참고해서 서비스 레이어에 구현하는게 좋아 보입니다.

추가로 createReservation 내 중복된 로직이 보이니 수정이 필요할 것 같습니다.

J-Cheol replied:
감사합니다. 수정완료

J-Cheol committed:
refactor : 검증 서비스 개편, 중복 로직 제거, 레이어에 맞는 기능 구현 등

예약관련 리팩토링
타도메인 리포지토리 사용 지양
재사용성을 위한 메서드 캡슐화
코드 분리

ppusda commented on src/main/java/com/toucheese/reservation/controller/ReservationController.java:
```diff
@@ -0,0 +1,76 @@
+	private final ReservationService reservationService;
+	private final ProductService productService;
+
+	@PostMapping("")
```
`("")`는 생략해도 되어서 빼도 될 것 같습니다.

J-Cheol replied:
네 확인했습니다!

J-Cheol committed:
refactor : 중복 로직 제거

J-Cheol committed:
refactor : 반환할때 request를 하지 않고 요청만 전달

ppusda commented on src/main/java/com/toucheese/product/service/ProductService.java:
```diff
@@ -28,4 +33,19 @@ public ProductDetailResponse findProductDetailById(Long productId) {
         return ProductDetailResponse.of(product);
     }
 
+    public Product findProductById(Long productId) {
```
새로 추가해주신 findBy 메서드들에 `@Transactional(readonly = true)`를 명시해주시면 좋을 것 같습니다.

J-Cheol replied:
추가하였습니다.

ppusda commented on src/main/java/com/toucheese/reservation/dto/ReservationRequest.java:
```diff
@@ -0,0 +1,44 @@
+	String name,
+	String phone,
+	LocalDate createDate,
+	LocalTime createTime,
```
`@Valid`를 사용하셨으니 입력 값에 대한 검증을 진행해야하는데, 해당 내용이 없습니다.
`@NotBlank`와 같이 해당하는 옵션과 메시지를 추가해주시면 될 것 같습니다.

위 내용에서는 아마 날짜에 대한 내용만 필수적으로 받는 걸 검증해보면 될 것 같은데 검토해주시고 적용해주시면 좋을 것 같습니다.

아래는 제가 로그인 기능 구현 시 사용한 예제입니다!

```java
@Builder
public record LoginRequest(
        @NotBlank(message = "아이디는 필수 값입니다.")
        String username,
        @NotBlank(message = "비밀번호는 필수 값입니다.")
        String password
) {

}
```

J-Cheol replied:
추가하였습니다~!

ppusda commented on src/main/java/com/toucheese/reservation/service/ReservationService.java:
```diff
@@ -0,0 +1,52 @@
+	private final ReservationValidationService reservationValidationService;
+
+	@Transactional
+	public Reservation createReservation(ReservationRequest reservationRequest) {
```
Controller에서 반환 값을 제외 했으니 Service 단의 반환 타입과 값을 조정해주시면 될 것 같습니다!

J-Cheol replied:
수정하였습니다

J-Cheol committed:
refactor : 불필요한 로직, 정보 요청 삭제 및 매개변수명 변경

ppusda commented on src/main/java/com/toucheese/reservation/service/ReservationService.java:
```diff
@@ -0,0 +1,59 @@
+
+		List<ReservationProductAddOption> reservationProductAddOptions = reservationRequest.addOptions().stream()
+			.map(addOptionRequest -> {
+				ProductAddOption productAddOption = productService.findProductAddOptionById(addOptionRequest.id());
```
로직이 어색한 것 같습니다.
`findProductAddOptionById`면 `ProductAddOptionId`로 검색을 시도해야하는데, 정작 인자 값으로 넣는 것은 `AddOptionId`인 것 같습니다.
제대로 조회가 되려면 `findProductAddOptionByProductIdAndAddOptionId` 와 같이 검색을 시도하여 조회된 단 하나의 `ProductAddOption`을 사용해야할 것 같습니다.

J-Cheol replied:
변수명을 잘못지었네요 ㅠㅠ 수정하였습니다.

J-Cheol replied:
잘못 이해한 점 수정하였습니다!

J-Cheol committed:
refactor : NotBlank 생성, 서비스 반환인자 타입 void, readOnly 추가

J-Cheol committed:
feat : addOptionId와 productId를 이용하여 해당하는 productAddOptionId 조회

productId 참조를 위하여 product와 productAddOption 양방향 관계 설정하였습니다.

addOptionId와 productId를 통해 해당하는 productAddOptionId를 저장합니다.

ppusda reviewed (approved):
수고하셨습니다!
